cpp/src/arrow/compute/util.cc

```diff
@@ -32,10 +32,10 @@ using internal::CpuInfo;
 namespace util {
 
 void TempVectorStack::alloc(uint32_t num_bytes, uint8_t** data, int* id) {
-  int64_t new_top = top_ + PaddedAllocationSize(num_bytes) + 2 * sizeof(uint64_t);
-  // Stack overflow check (see GH-39582).
+  const auto estimate_size = EstimateAllocSize(num_bytes);
   // XXX cannot return a regular Status because most consumers do not either.
-  ARROW_CHECK_LE(new_top, buffer_size_) << "TempVectorStack::alloc overflow";
+  ARROW_DCHECK_OK(CheckAllocOverflow(estimate_size));
```

Comment:
We probably should use `ARROW_CHECK_OK` here?

Reply:
Yes, you're right. ARROW_DCHECK_OK seems not work in NDEBUG mode.